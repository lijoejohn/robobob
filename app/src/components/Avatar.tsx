import clsx from "clsx";
export const Avatar = ({ isQuestion }: { isQuestion?: boolean }) => {
  return (
    <div
      className={clsx(
        "rounded",
        "p-6",
        "flex",
        "justify-center",
        "items-center",
        "w-[20px]",
        "h-[20px]",
        !isQuestion && "mr-4",
        isQuestion && "ml-4",
        isQuestion && "bg-white",
        isQuestion && "text-black",
        !isQuestion && "bg-red-900",
        !isQuestion && "text-white"
      )}
    >
      {isQuestion ? `You` : `Bob`}
    </div>
  );
};
