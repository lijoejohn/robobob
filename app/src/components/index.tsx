export * from "./MessageBlock";
export * from "./TextArea";
export * from "./Avatar";
export * from "./icons";
export * from "./Buttons";
export * from "./ErrorBoundary";
