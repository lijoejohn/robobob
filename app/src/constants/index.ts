export const API_ENDPOINT = `https://frozen-sea-84259-a7779bc76c91.herokuapp.com`;

export const scrollStyle =
  window.innerWidth > 768
    ? { maxHeight: "calc(100vh - 40px)" }
    : {
        maxHeight: "calc(100vh - 150px)",
      };
