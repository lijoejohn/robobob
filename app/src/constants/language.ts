export const LABELS = {
  ERROR_TEXT: "Something went wrong, Please reload the page",
  APP_NAME: "Robo - Bob",
  RECENT_QUESTIONS: "What you requested of me",
  YOU: "You",
  BOB: "Bob",
  BASIC_MATH_EXPRESSIONS: "Basic mathematical expressions",
  BASIC_MATH_EXPRESSION_ANSWER:
    "Your response to the basic mathematical expressions",
  ANSWER_NOT_FOUD:
    "I'm sorry, but I have no knowledge regarding this query. Would you kindly reword the questions?",
};
