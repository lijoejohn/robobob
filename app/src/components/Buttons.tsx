import { clsx } from "clsx";
import { Icon } from "../components";

interface ExtraProps {
  icon: string;
}

export const IconButton = (
  props: React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  > &
    ExtraProps
) => {
  return (
    <button
      className={clsx(
        "h-12",
        "px-5",
        "py-2",
        "border",
        "border-gray-400",
        "border-solid",
        "divide-transparent",
        "rounded-md",
        "hover:border-white",
        "focus-within:outline-blue-200",
        "focus:outline-4",
        "focus:outline-blue-200",
        "active:outline-blue-200",
        "disabled:bg-gray-500",
        "disabled:border-gray-400",
        " disabled:cursor-not-allowed"
      )}
      {...props}
    >
      <Icon icon={props.icon} />
    </button>
  );
};

export const SecondaryButton = (
  props: React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  >
) => {
  return (
    <button
      className={clsx(
        "border-gray-400",
        "border",
        "divide-transparent",
        "border-solid",
        "truncate",
        "m-2",
        "max-w-[14rem]",
        "rounded",
        "p-0",
        "text-xs",
        "font-light",
        "mr-2",
        "px-2.5",
        "py-0.5",
        "hover:border-white",
        "focus-within:outline-blue-200",
        "focus:outline-4",
        "focus:outline-blue-200"
      )}
      {...props}
    >
      {props.children}
    </button>
  );
};
