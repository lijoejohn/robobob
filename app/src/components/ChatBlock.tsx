import { MessageBlock } from "../components";
import { ThreadType, Thread } from "../annotations";

export const ChatBlock = ({ threadsState }: { threadsState: Thread[] }) => {
  return threadsState.map((dataSet) => {
    return (
      <MessageBlock
        key={dataSet.threadKey}
        isQuestion={dataSet.threadType === ThreadType.Question}
        thread={dataSet.thread}
      />
    );
  });
};
