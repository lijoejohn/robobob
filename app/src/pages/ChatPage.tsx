import { useEffect, useState, useRef } from "react";
import SimpleBar from "simplebar-react";
import { TextArea, IconButton, ChatBlock } from "../components";
import { useQuestionsState } from "../global-state/QuestionsContext";
import { useThreadsState } from "../global-state/ThreadsContext";
import { AnswerResponseType } from "../annotations";
import "simplebar-react/dist/simplebar.min.css";
import {
  isMathExpression,
  resetRecentQuestions,
  mathsQuestionThread,
  mathsAnswerThread,
  questionThread,
  unknownAnswerThread,
  answerThread,
} from "../helpers";
import UseFetch from "../hooks/useFetch";
import { scrollStyle } from "../config";

const ChatPage = ({ recentQuestion }: { recentQuestion: string }) => {
  const ref = useRef<HTMLElement>();
  const inputBoxRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLTextAreaElement>(null);

  const [question, setQuestion] = useState<string>("");
  const [activeQuestion, setActiveQuestion] = useState<string>("");
  const [loading, setLoading] = useState<boolean>(false);

  const { questionsState, setQuestionsState } = useQuestionsState();
  const { threadsState, setThreadsState } = useThreadsState();

  useEffect(() => {
    ref?.current?.scrollTo({ top: inputBoxRef.current?.offsetTop });
    inputRef.current?.focus();
  }, []);

  useEffect(() => {
    ref?.current?.scrollTo({
      top: inputBoxRef.current?.offsetTop! + 500,
    });
  }, [threadsState]);

  useEffect(() => {
    setQuestion(recentQuestion);
  }, [recentQuestion]);

  const typeHandler = (e: React.FormEvent<HTMLTextAreaElement>) => {
    const newValue = e.currentTarget.value;
    setQuestion(newValue);
  };

  const submitHandler = () => {
    const recentQuestions = resetRecentQuestions(questionsState, question);

    setQuestionsState(recentQuestions);
    if (isMathExpression(question)) {
      const questionThread = mathsQuestionThread(question);
      const answerThread = mathsAnswerThread(question);
      setThreadsState([...threadsState, questionThread, answerThread]);
      setQuestion("");
    } else {
      setLoading(true);
      setActiveQuestion(question);
    }
  };

  return (
    <div className="absolute mt-5 bottom-0 w-[calc(100%-5rem)] sm:w-[calc(100%-20rem)] md:w-[calc(100%-20rem)]">
      <SimpleBar
        id="scroll"
        scrollableNodeProps={{ ref: ref }}
        style={scrollStyle}
      >
        <ChatBlock threadsState={threadsState} />

        <div className="flex items-center p-5 mb-6" ref={inputBoxRef}>
          <TextArea
            ref={inputRef}
            value={question}
            placeholder="Type a thread"
            id="first_name"
            onChange={typeHandler}
            onKeyDown={(e) => {
              if (e.key === "Enter" && e.shiftKey == false) {
                e.preventDefault();
                submitHandler();
              }
            }}
            disabled={loading}
          />
          <IconButton
            icon="sendMessage"
            aria-label="Send Message"
            onClick={submitHandler}
            disabled={question === "" || loading}
          />
        </div>
      </SimpleBar>
      {loading ? (
        <UseFetch
          callBack={(thread: AnswerResponseType | undefined) => {
            setLoading(false);
            setQuestion("");
            const questionSet = questionThread(question);
            if (!thread) {
              const answerThread = unknownAnswerThread();
              setThreadsState([...threadsState, questionSet, answerThread]);
            } else {
              const answerSet = answerThread(thread.answer);
              setThreadsState([...threadsState, questionSet, answerSet]);
            }
          }}
          thread={activeQuestion}
        />
      ) : null}
    </div>
  );
};
export default ChatPage;
