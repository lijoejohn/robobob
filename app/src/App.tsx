import { useState } from "react";
import ChatPage from "./pages/ChatPage";
import { SideBar } from "./components/layout/";
import { ErrorBoundary } from "./components";
import { QuestionsProvider } from "./global-state/QuestionsContext";
import { ThreadsProvider } from "./global-state/ThreadsContext";

function App() {
  const [recentQuestion, setRecentQuestion] = useState<string>("");
  return (
    <ErrorBoundary>
      {/* Recent questions context provider */}
      <QuestionsProvider>
        <SideBar setRecentQuestion={setRecentQuestion} />
        <div className="h-screen p-4 sm:ml-64 w-content">
          <div className="h-full p-4 border-2 border-gray-700 border-solid rounded-lg">
            {/* Chat conversation thread provider */}
            <ThreadsProvider>
              <ChatPage recentQuestion={recentQuestion} />
            </ThreadsProvider>
          </div>
        </div>
      </QuestionsProvider>
    </ErrorBoundary>
  );
}

export default App;
