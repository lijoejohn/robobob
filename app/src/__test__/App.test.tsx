import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";
import * as renderer from "react-test-renderer";

import App from "../App";
import { SideBar, Header } from "../components/layout";
import ChatPage from "../pages/ChatPage";
import { ErrorBoundary } from "../components";
import { LABELS } from "../constants/language";

const ErrorPage = () => {
  throw new Error();
};

describe("Test APP Snapshot", function () {
  it("Matches snapshot", () => {
    const app = renderer.create(<App />).toJSON();
    expect(app).toMatchSnapshot();
  });
});

describe("Test App Accessibility", function () {
  it("Should render with Standard Accessibility", async function () {
    expect.extend(toHaveNoViolations);
    const { container } = render(<App />);
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("Test App Semantic", function () {
  test("Should render app with Header,Main and Footer", () => {
    const testRenderer = renderer.create(<App />);
    const testInstance = testRenderer.root;
    expect(testInstance.findByType(Header));
    expect(testInstance.findByType(SideBar));
    expect(testInstance.findByType(ChatPage));
  });
});

describe("<ErrorBoundary>", () => {
  it("displays error message on error generated by child", () => {
    render(
      <ErrorBoundary>
        <ErrorPage />
      </ErrorBoundary>
    );
    expect(screen.getByRole("heading")).toHaveTextContent(LABELS.ERROR_TEXT);
  });
  it("should not displays error message on valid child", () => {
    render(
      <ErrorBoundary>
        <h1>Hello</h1>
      </ErrorBoundary>
    );
    expect(screen.getByRole("heading")).toHaveTextContent("Hello");
  });
});
