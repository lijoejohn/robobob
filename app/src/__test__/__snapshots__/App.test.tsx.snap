// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test APP Snapshot Matches snapshot 1`] = `
[
  <aside
    aria-label="Sidebar"
    className="fixed top-0 left-0 z-40 w-64 h-screen transition-transform -translate-x-full sm:translate-x-0"
    id="default-sidebar"
  >
    <div
      className="h-full overflow-y-auto bg-brandPrimary"
    >
      <div
        className="mt-6"
      >
        <div
          className="justify-center mb-4"
        >
          <header
            className="p-0 m-0 text-lg font-bold"
            data-testid="app-name"
          >
            Robo - Bob
          </header>
        </div>
        <hr
          className="border-t-0.5"
        />
      </div>
    </div>
  </aside>,
  <div
    className="h-screen p-4 sm:ml-64 w-content"
  >
    <div
      className="h-full p-4 border-2 border-gray-700 border-solid rounded-lg"
    >
      <div
        className="absolute mt-5 bottom-0 w-[calc(100%-5rem)] sm:w-[calc(100%-20rem)] md:w-[calc(100%-20rem)]"
      >
        <div
          data-simplebar="init"
          id="scroll"
          style={
            {
              "maxHeight": "calc(100vh - 40px)",
            }
          }
        >
          <div
            className="simplebar-wrapper"
          >
            <div
              className="simplebar-height-auto-observer-wrapper"
            >
              <div
                className="simplebar-height-auto-observer"
              />
            </div>
            <div
              className="simplebar-mask"
            >
              <div
                className="simplebar-offset"
              >
                <div
                  aria-label="scrollable content"
                  className="simplebar-content-wrapper"
                  role="region"
                  tabIndex={0}
                >
                  <div
                    className="simplebar-content"
                  >
                    <div
                      className="flex items-center p-5 mb-6"
                    >
                      <textarea
                        className="w-full h-12 mr-5 border text-sm rounded-lg p-2.5 bg-gray-700 border-gray-700 focus:border-blue-700"
                        data-testid="send-message"
                        disabled={false}
                        onChange={[Function]}
                        onKeyDown={[Function]}
                        placeholder="Type a thread"
                        value=""
                      />
                      <button
                        aria-label="Send Message"
                        className="h-12 px-5 py-2 border border-gray-400 border-solid divide-transparent rounded-md hover:border-white focus-within:outline-blue-200 focus:outline-4 focus:outline-blue-200 active:outline-blue-200 disabled:bg-gray-500 disabled:border-gray-400  disabled:cursor-not-allowed"
                        data-testid="send-button"
                        disabled={true}
                        icon="sendMessage"
                        onClick={[Function]}
                      >
                        <svg
                          className="w-12 h-4 m-1 md:m-0"
                          fill="none"
                          strokeWidth="2"
                          viewBox="0 0 12 16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M.5 1.163A1 1 0 0 1 1.97.28l12.868 6.837a1 1 0 0 1 0 1.766L1.969 15.72A1 1 0 0 1 .5 14.836V10.33a1 1 0 0 1 .816-.983L8.5 8 1.316 6.653A1 1 0 0 1 .5 5.67V1.163Z"
                            fill="currentColor"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="simplebar-placeholder"
            />
          </div>
          <div
            className="simplebar-track simplebar-horizontal"
          >
            <div
              className="simplebar-scrollbar"
            />
          </div>
          <div
            className="simplebar-track simplebar-vertical"
          >
            <div
              className="simplebar-scrollbar"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
